File: typescript/hackerrank/cutTheSticks.ts
// https://www.hackerrank.com/challenges/cut-the-sticks/problem

function cutTheSticks_IterativeSolution(arr: number[]): number[] {
    const result: number[] = [];
    let inputArr = [...arr];
    while (inputArr.length > 0) {
        const min = Math.min(...inputArr)
        if (min != null) {
             result.push(inputArr.length);
             inputArr = inputArr.filter(i => i != min).map(i => i - min);
         }
    }
     return result;
}

function cutTheSticks_RecursiveSolution(arr: number[]): number[] {
    const result = _cutArray(arr)
    return result;
}

function _cutArray(input: number[], output: number[] = []): number[] {
    if (input.length == 0) return output
    let inputArr = [...input];
    const min = Math.min(...inputArr);
    if (min != null) {
        output.push(inputArr.length)
        inputArr = inputArr.filter(i => i != min).map(i => i - min);
    }
    return _cutArray(inputArr, output);
}


test("sample input 0", () => {
    const arr = [5, 4, 4, 2, 2, 8], expected = [6, 4, 2, 1];
    const result2 = cutTheSticks_RecursiveSolution(arr);
    const result1 = cutTheSticks_IterativeSolution(arr);
    expect(result2).toEqual(expected);
    expect(result1).toEqual(expected);
});

test("sample input 1", () => {
    const arr = [1, 2, 3, 4, 3, 3, 2, 1], expected = [8, 6, 4, 1];
    const result2 = cutTheSticks_RecursiveSolution(arr);
    const result1 = cutTheSticks_IterativeSolution(arr);
    expect(result2).toEqual(expected);
    expect(result1).toEqual(result2);
});

File: kotlin/src/hackerrank/cutTheSticks.kt
package hackerrank

import org.junit.jupiter.api.Assertions.assertArrayEquals
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test

// https://www.hackerrank.com/challenges/cut-the-sticks/problem

fun cutTheSticks_IterativeSolution(arr: Array<Int>): Array<Int> {
    val result = mutableListOf<Int>()
    var inputArr : Array<Int> = arr.clone()
    while (inputArr.isNotEmpty()) {
        val min = inputArr.min()
        if (min != null) {
            result.add(inputArr.size)
            inputArr = inputArr.filter { it: Int -> it != min }
                            .map { it: Int -> it - min }
                            .toTypedArray()
        }
    }
    return result.toTypedArray()
}

fun cutTheSticks_RecursiveSolution(arr: Array<Int>): Array<Int> {
    var result = mutableListOf<Int>()
    result = cutArray(arr, result)
    return result.toTypedArray()
}

private fun cutArray(input: Array<Int>, output: MutableList<Int>): MutableList<Int> {
    if (input.isEmpty()) return output
    var inputArr = input.clone()
    val min = inputArr.min()
    if (min != null) {
        output.add(inputArr.size)
        inputArr = inputArr.filter { it: Int -> it != min }
                    .map { it: Int -> it - min }
                    .toTypedArray()
    }
    return cutArray(inputArr, output)
}


internal class CutTheSticksTest {

    @Test
    @DisplayName("Test sample input 0")
    fun cutTheSticks0() {
        val arr = arrayOf<Int>(5, 4, 4, 2, 2, 8)
        val expected = arrayOf<Int>(6, 4, 2, 1)
        val result2 = cutTheSticks_RecursiveSolution(arr)
        val result1 = cutTheSticks_IterativeSolution(arr)
        assertArrayEquals(expected, result2)
        assertArrayEquals(expected, result1)
    }

    @Test
    @DisplayName("Test sample input 1")
    fun cutTheSticks1() {
        val arr = arrayOf(1, 2, 3, 4, 3, 3, 2, 1)
        val expected = arrayOf(8, 6, 4, 1)
        val result2 = cutTheSticks_RecursiveSolution(arr)
        val result1 = cutTheSticks_IterativeSolution(arr)
        assertArrayEquals(expected, result2)
        assertArrayEquals(expected, result1)
    }

}